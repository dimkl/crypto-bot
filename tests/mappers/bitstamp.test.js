const BitstampMapper = require('../../src/mappers/bitstamp');

describe('BitstampMapper(options)', () => {
  beforeEach(() => {
    this.mapper = new BitstampMapper({ currencyPair: 'xlmeur' });
  });

  describe('constructor(options)', () => {
    test('currencyPair', () => {
      expect(this.mapper.currencyPair).toBe('xlmeur');
    });
  });

  describe('properties', () => {
    test('availableKeys', () => {
      expect(this.mapper.availableKeys).toEqual([
        'xlm_available',
        'eur_available'
      ]);
    });

    test('feeKey', () => {
      expect(this.mapper.feeKey).toBe('xlmeur_fee');
    });

    test('exchangeRateKey', () => {
      expect(this.mapper.exchangeRateKey).toBe('xlm_eur');
    });
  });

  describe('_getTransactionType(type)', () => {
    test('deposit', () => {
      expect(this.mapper._getTransactionType(0)).toBe('deposit');
    });

    test('withdrawl', () => {
      expect(this.mapper._getTransactionType(1)).toBe('withdrawl');
    });

    test('market_trade', () => {
      expect(this.mapper._getTransactionType(2)).toBe('market_trade');
    });

    test('sub_account_transfer', () => {
      expect(this.mapper._getTransactionType(14)).toBe('sub_account_transfer');
    });
  });

  describe('_getExchangeType(capital)', () => {
    test('buy', () => {
      expect(this.mapper._getExchangeType(0)).toBe('buy');
    });

    test('sell', () => {
      expect(this.mapper._getExchangeType(100)).toBe('sell');
    });
  });

  describe('liveValues(data)', () => {
    test('returns mapped data', () => {
      const data = {
        open: '0.33391300',
        bid: '0.31954600',
        ask: '0.32000900',
        vwap: '0.33508642'
      };

      expect(this.mapper.liveValues(data)).toMatchObject({
        open: '0.33391300',
        currentBid: '0.31954600',
        currentAsk: '0.32000900',
        vwap: '0.33508642'
      });
    });
  });

  describe('hourlyValues(data)', () => {
    test('returns mapped data', () => {
      const data = {
        open: '0.33391300',
        bid: '0.31954600',
        ask: '0.32000900',
        vwap: '0.33508642'
      };

      expect(this.mapper.hourlyValues(data)).toMatchObject({
        hourlyBid: '0.31954600',
        hourlyAsk: '0.32000900',
        hourlyOpen: '0.33391300',
        hourlyVwap: '0.33508642'
      });
    });
  });

  describe('accountBalance(data)', () => {
    test('returns mapped data', () => {
      const data = {
        eur_available: '20.00000000',
        xlm_available: '960.31161000',
        xlmeur_fee: 0.5
      };
      expect(this.mapper.accountBalance(data)).toMatchObject({
        assets: '960.31161000',
        capital: '20.00000000',
        feePercentage: '0.0050'
      });
    });
  });

  describe('sell(data)', () => {
    test('returns mapped data', () => {
      const now = Date.now();

      const data = {
        price: '0.60000',
        amount: '20.00000000',
        datetime: now,
        id: 'O4T556-LSPCF-JNWZPH'
      };

      expect(this.mapper.sell(data)).toMatchObject({
        orderId: 'O4T556-LSPCF-JNWZPH',
        soldAt: now,
        soldValue: '0.60000',
        soldAmount: '20.00000000'
      })
    })
  });

  describe('buy(data)', () => {
    test('returns mapped data', () => {
      const now = Date.now();

      const data = {
        price: '0.60000',
        amount: '20.00000000',
        datetime: now,
        id: 'O4T556-LSPCF-JNWZPH'
      };

      expect(this.mapper.buy(data)).toMatchObject({
        orderId: 'O4T556-LSPCF-JNWZPH',
        boughtAt: now,
        boughtValue: '0.60000',
        boughtAmount: '20.00000000'
      })
    })
  });
});