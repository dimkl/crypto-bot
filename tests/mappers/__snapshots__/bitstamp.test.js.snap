// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitstampMapper(options) userTransactions(data) returns mapped data 1`] = `
Array [
  Object {
    "assets": 0.0000048,
    "capital": 0,
    "datetime": "2021-02-13 02:03:57.546000",
    "exchangeRate": 0.477,
    "exchangeType": "sell",
    "feeAmount": "0.00000",
    "orderId": 1328524713971714,
    "transactionId": 150706592,
    "transactionType": "market_trade",
  },
  Object {
    "assets": 461.0838,
    "capital": 219.93697,
    "datetime": "2021-02-13 02:03:49.523000",
    "exchangeRate": 0.477,
    "exchangeType": "sell",
    "feeAmount": "1.09968",
    "orderId": 1328524713971714,
    "transactionId": 150706570,
    "transactionType": "market_trade",
  },
]
`;
